src/main/java/com/google/api/codegen/transformer/ApiMethodTransformer.java

```diff
@@ -0,0 +1,260 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.transformer;
+
+import com.google.api.codegen.MethodConfig;
+import com.google.api.codegen.ServiceMessages;
+import com.google.api.codegen.util.Name;
+import com.google.api.codegen.viewmodel.ApiMethodDocView;
+import com.google.api.codegen.viewmodel.ApiMethodType;
+import com.google.api.codegen.viewmodel.DynamicDefaultableParamView;
+import com.google.api.codegen.viewmodel.MapParamDocView;
+import com.google.api.codegen.viewmodel.OptionalArrayMethodView;
+import com.google.api.codegen.viewmodel.ParamDocView;
+import com.google.api.codegen.viewmodel.RequestObjectParamView;
+import com.google.api.codegen.viewmodel.SimpleParamDocView;
+import com.google.api.tools.framework.model.Field;
+import com.google.api.tools.framework.model.TypeRef;
+import com.google.protobuf.DescriptorProtos.FieldDescriptorProto.Type;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * ApiMethodTransformer generates view objects from method definitions.
+ */
+public class ApiMethodTransformer {
+  private InitCodeTransformer initCodeTransformer;
+
+  public ApiMethodTransformer() {
+    this.initCodeTransformer = new InitCodeTransformer();
+  }
+
+  public OptionalArrayMethodView generateOptionalArrayMethod(MethodTransformerContext context) {
+    SurfaceNamer namer = context.getNamer();
+    OptionalArrayMethodView.Builder apiMethod = OptionalArrayMethodView.newBuilder();
+
+    if (context.getMethodConfig().isPageStreaming()) {
+      apiMethod.type(ApiMethodType.PagedOptionalArrayMethod);
+    } else {
+      apiMethod.type(ApiMethodType.OptionalArrayMethod);
+    }
+    apiMethod.apiClassName(namer.getApiWrapperClassName(context.getInterface()));
+    apiMethod.apiVariableName(namer.getApiWrapperVariableName(context.getInterface()));
+    apiMethod.initCode(
+        initCodeTransformer.generateInitCode(
+            context, context.getMethodConfig().getRequiredFields()));
+
+    ApiMethodDocView.Builder docBuilder = ApiMethodDocView.newBuilder();
+
+    docBuilder.mainDocLines(namer.getDocLines(context.getMethod()));
+    List<ParamDocView> paramDocs =
+        getMethodParamDocs(context, context.getMethodConfig().getRequiredFields());
+    paramDocs.add(getOptionalArrayParamDoc(context, context.getMethodConfig().getOptionalFields()));
+    paramDocs.add(getCallSettingsParamDoc(context));
+    docBuilder.paramDocs(paramDocs);
+    docBuilder.returnTypeName(
+        namer.getDynamicReturnTypeName(
+            context.getTypeTable(), context.getMethod(), context.getMethodConfig()));
+    docBuilder.throwsDocLines(new ArrayList<String>());
+
+    apiMethod.doc(docBuilder.build());
```

Comment:
Perhaps put docBuilder creation in its own method? And the same with methodParams?